plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/BuildPath.java

```diff
@@ -79,53 +73,66 @@ public static void addMavenLibraries(IProject project, List<Library> libraries,
   }
 
   /**
-   * @return the entries added to the classpath.
-   *     Does not include entries previously present in classpath.
+   * Returns the entries added to the classpath.
    */
-  static IClasspathEntry[] addLibraries(
-      IJavaProject javaProject, List<Library> libraries, IProgressMonitor monitor)
-          throws JavaModelException, CoreException {
-
-    // todo or we could grab the transitive dependency graph here if needed
-    // at this point we have the Java project so we could even filter
-    // entries already present in other libraries
+  public static IClasspathEntry[] addNativeLibrary(IJavaProject javaProject,
+      List<Library> libraries, IProgressMonitor monitor) throws CoreException {
     
-    return prepareLibraries(javaProject, libraries, monitor, true);
-  }
-
-  private static IClasspathEntry[] prepareLibraries(IJavaProject javaProject,
-      List<Library> libraries, IProgressMonitor monitor, boolean addToClasspath)
-          throws JavaModelException, CoreException {
     SubMonitor subMonitor = SubMonitor.convert(monitor,
         Messages.getString("adding.app.engine.libraries"), //$NON-NLS-1$
-        libraries.size() + 1); // + 1 because we pass the submonitor along below
-
-    List<IClasspathEntry> rawClasspath = Lists.newArrayList(javaProject.getRawClasspath());
+        2);
+    
+    Library masterLibrary = collectLibraryFiles(javaProject, libraries);
+    
+    List<IClasspathEntry> newEntries = computeEntries(javaProject, masterLibrary);
```

Comment:
As with the `addNativeLibrary` naming: if we make this class only for the master library, `computeEntries` always return either 1) an empty list if `javaProject` already has the master container; or 2) a singleton list containing the master container. So, it would be much less confusing to do

```java
   if (javaProject doesn't have the master container) {
       ClasspathUtil.addClasspathEntries(..., newEntries /* a one-size array containing the master container */, ...);
   }
   runContainerResolverJob();
}
```
As I said, we could do this in a separate PR.

Reply:
Again, let's wait to see where we land once we load the dependencies from Maven, and don't have library dependencies.